### PR TITLE
SDV2: add safety_checker

### PR DESCRIPTION
**Pull request by Songyuanwei**

### CODE REVIEW

File: examples/stable_diffusion_v2/tools/safety_checker/nsfw_model.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+
+class Normalization(nn.Cell):
+    def __init__(self, shape):
+        super().__init__()
+        self.mean = Parameter(Tensor(np.zeros(shape), ms.float32))
+        self.variance = Parameter(Tensor(np.ones(shape), ms.float32))
+
+    def construct(self, x):
+        return (x - self.mean) / ms.ops.Sqrt()(self.variance)
```
Comment (zhtmike) [inline examples/stable_diffusion_v2/tools/safety_checker/nsfw_model.py:14]: maybe use `ops.sqrt` ?

File: examples/stable_diffusion_v2/tools/safety_checker/safety_checker.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+            self.dtype = ms.float32 if config.dtype == "float32" else ms.float16
+            model = CLIPModel(config)
+            processor = CLIPImageProcessor()
+            tokenizer = CLIPTokenizer(tokenizer_path, pad_token="!")
+
+            def process_text(text):
+                return ms.Tensor(tokenizer(text, padding="max_length", max_length=77)["input_ids"])
+
+            if safety_version == 2:
+                from tools.safety_checker.nsfw_model import NSFWModel
```
Comment (zhtmike) [inline examples/stable_diffusion_v2/tools/safety_checker/safety_checker.py:72]: suggest to put `import` things on the top of the file. May not be a good practice to put them in the content

File: examples/stable_diffusion_v2/tools/safety_checker/safety_checker.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+            results.append(result_img)
+        return results
+
+    def cosine_distance(self, v, w):
+        if self.backend == "ms":
+            v /= L2_norm_ops(v)
+            w /= L2_norm_ops(w)
+            return ops.matmul(v, w.T)
+        else:
+            import torch
```
Comment (zhtmike) [inline examples/stable_diffusion_v2/tools/safety_checker/safety_checker.py:145]: same, we can add a `try: import torch except ImportError: is_torch_available=False` on top of the file. Raise RuntimeError when using torch backend while Torch is not installed

File: examples/stable_diffusion_v2/tools/eval/README.md [hunk 4]
```diff
@@ -9,7 +9,7 @@ This folder contains code and scripts for diffusion model evaluation, e.g.,
 
 Note that all the above metrics are computed based on neural network models.
 
-> A convincing evaluation for diffusion models requires both visually qualitative comparision and quantitative measure. A higher FID score (or CLIP score) does not nessarily show one model is better than another.
+> A convincing evaluation for diffusion models requires both visually qualitative comparision and quantitative measure. A lower FID score (or a higher CLIP score) does not nessarily show one model is better than another.
```
Comment (wtomin) [inline examples/stable_diffusion_v2/tools/eval/README.md:12]: typo: nessarily ->  necessarily

File: examples/stable_diffusion_v2/tools/safety_checker/safety_check.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
+# Safety Checker
+
+## Introduction
+To validate the safeness of images generated by a model, we implement two safety checkers in `safety_checker.py` with `safety_version=1` and `safety_version=2`(Default).
+
+`safety_version=1` applies CLIP to compute the similarity between the given images and a pre-defined list of NSFW concepts (two possible `.yaml` files to choose from). This is consistent with the one used in CompVis's stable diffusion 1.x from `diffusers` implementation. The output is a list of similarity scores, each corresponding to a concept.
+
+`safety_version=2` trains a NSFW classifier with a supervisied approach, taking image features from CLIP as its input. This is consistent with the one used in StabilityAI's stable diffusion 2.x. The output is a number in 0-1, representing the probability of the generated image being NSFW.
```
Comment (wtomin) [inline examples/stable_diffusion_v2/tools/safety_checker/safety_check.md:8]: typo: supervisied -> supervised.

File: examples/stable_diffusion_v2/tools/safety_checker/nsfw_model.py (hunk 1 above)
Comment (Songyuanwei) [reply examples/stable_diffusion_v2/tools/safety_checker/nsfw_model.py]: fixed

File: examples/stable_diffusion_v2/tools/safety_checker/safety_checker.py (hunk 2 above)
Comment (Songyuanwei) [reply examples/stable_diffusion_v2/tools/safety_checker/safety_checker.py]: fixed

File: examples/stable_diffusion_v2/tools/safety_checker/safety_checker.py (hunk 3 above)
Comment (Songyuanwei) [reply examples/stable_diffusion_v2/tools/safety_checker/safety_checker.py]: fixed

File: examples/stable_diffusion_v2/tools/eval/README.md (hunk 4 above)
Comment (Songyuanwei) [reply examples/stable_diffusion_v2/tools/eval/README.md]: fixed

File: examples/stable_diffusion_v2/tools/safety_checker/safety_check.md (hunk 5 above)
Comment (Songyuanwei) [reply examples/stable_diffusion_v2/tools/safety_checker/safety_check.md]: fixed

**Review by zhtmike — APPROVED** [review]

**Review by wtomin — APPROVED** [review]

File: examples/stable_diffusion_v2/tools/safety_checker/nsfw_model.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+import mindspore as ms
+from mindspore import Parameter, Tensor, nn, ops
+
+
+class Normalization(nn.Cell):
+    def __init__(self, shape):
+        super().__init__()
+        self.mean = Parameter(Tensor(np.zeros(shape), ms.float32))
+        self.variance = Parameter(Tensor(np.ones(shape), ms.float32))
```
Comment (hadipash) [inline examples/stable_diffusion_v2/tools/safety_checker/nsfw_model.py:11]: PyTorch's `register_buffer` stores non-trainable states. So, you can set `requires_grad=False` in `Parameter` or use `Tensor`.

File: examples/stable_diffusion_v2/tools/safety_checker/safety_checker.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+import argparse
+import os
+import sys
+
+import numpy as np
+import yaml
+from PIL import Image
+
+# equivalent to no-check-certificate flag in wget
+os.environ["CURL_CA_BUNDLE"] = ""
```
Comment (hadipash) [inline examples/stable_diffusion_v2/tools/safety_checker/safety_checker.py:14]: It is not safe to disable certificate verification for everyone. Perhaps you can describe it as an option in `safety_check.md`?

File: examples/stable_diffusion_v2/tools/safety_checker/nsfw_model.py (hunk 6 above)
Comment (Songyuanwei) [reply examples/stable_diffusion_v2/tools/safety_checker/nsfw_model.py]: fixed

File: examples/stable_diffusion_v2/tools/safety_checker/safety_checker.py (hunk 7 above)
Comment (Songyuanwei) [reply examples/stable_diffusion_v2/tools/safety_checker/safety_checker.py]: fixed

**Review by hadipash — APPROVED** [review]

File: examples/stable_diffusion_v2/text_to_image.py [hunk 8]
```diff
@@ -468,6 +475,23 @@ def main(args):
...
     )
+    parser.add_argument(
+        "--check_safety",
+        action="store_true",
+        help="set this flag to use a safety checker",
+    )
+    parser.add_argument(
+        "--clip_ckpt_path",
+        type=str,
+        default=None,
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/text_to_image.py:486]: ckip ckpt path is already uploaded to the download server. It's better support auto download if `clip_ckpt_path` is None.

Comment (Songyuanwei) [reply examples/stable_diffusion_v2/text_to_image.py:487]: fixed

**Pull request merged**